src/databricks/sql/cloudfetch/download_manager.py

```diff
@@ -0,0 +1,139 @@
+from collections import namedtuple
+from concurrent.futures import ThreadPoolExecutor
+from databricks.sql.cloudfetch.downloader import ResultSetDownloadHandler
+
+
+class ResultFileDownloadManager:
+
+    def __init__(self, max_download_threads, lz4_compressed):
+        self.download_handlers = []
+        self.thread_pool = ThreadPoolExecutor(max_workers=max_download_threads + 1)
+        self.downloadable_result_settings = _get_downloadable_result_settings(lz4_compressed)
+        self.fetch_need_retry = False
+        self.num_consecutive_result_file_download_retries = 0
+        self.cloud_fetch_index = 0
+
+    def add_file_links(self, t_spark_arrow_result_links, next_row_index):
+        for t_spark_arrow_result_link in t_spark_arrow_result_links:
+            if t_spark_arrow_result_link.rowCount <= 0:
+                continue
+            self.download_handlers.append(ResultSetDownloadHandler(
+                self.downloadable_result_settings, t_spark_arrow_result_link))
+        self.cloud_fetch_index = next_row_index
+
+    def get_next_downloaded_file(self, next_row_index):
+        if not self.download_handlers:
+            return None
+
+        # Remove handlers we don't need anymore
+        self._remove_past_handlers(next_row_index)
+
+        # Schedule the downloads
+        self._schedule_downloads()
+
+        # Find next file
+        idx = self._find_next_file_index(next_row_index)
+        if idx is None:
+            return None
+        handler = self.download_handlers[idx]
+
+        # Check (and wait) for download status
+        if self._check_if_download_successful(handler):
+            # Buffer should be empty so set buffer to new ArrowQueue with result_file
+            result = DownloadedFile(
+                handler.result_file,
+                handler.result_link.startRowOffset,
+                handler.result_link.rowCount,
+            )
+            self.cloud_fetch_index += handler.result_link.rowCount
+            self.download_handlers.pop(idx)
+            # Return True upon successful download to continue loop and not force a retry
+            return result
+        # Download was not successful for next download item, force a retry
+        return None
+
+    def _remove_past_handlers(self, next_row_index):
+        """
+        Remove any download handlers whose start to end range doesn't include the next row to be fetched
+        i.e. no need to download
+        """
+        i = 0
```

Comment:
Isn't the point of cloudfetch to allow us to fetch results "out of order" in chunks and then re-assemble them once all downloads are complete? If I understand correctly, why do we need to expire handlers on the basis of the current row-count? Isn't that arbitrarily slowing the download process?

Reply:
Yes, we initiate the downloads of multiple cloud fetch files in parallel, but we read these files one by one in order to fill our resulting pyarrow table. So yes, we don't really need to expire handlers since we would be going in order and won't have any "past handlers". But, it's nice to keep in case we have actually moved past the row range of a handler